File: PNsAV/src/core/engine/engine.cpp
#include "engine.h"

void Engine::add_atom(const Atom& atom) {
    atoms.push_back(atom);
}

void Engine::add_rule(const Rule& rule) {
    rules.push_back(rule);
}

void Engine::add_argument(const Argument& argument) {
    arguments.push_back(argument);
}

void Engine::add_attack(const Attack& attack) {
    attacks.push_back(attack);
}

void Engine::build_attack_map() {
    for (const auto& attack : attacks) {
        attack_index[attack.attacker].push_back(attack.target);
        inv_attack_index[attack.target].push_back(attack.attacker);
    }
}

void Engine::build_argumentadj_vector() {
    for (const auto& arg : arguments) {
        ArgumentAdj arg_adj;
        arg_adj.id = arg.id;
        arg_adj.conclusion = arg.conclusion;
        arg_adj.top_rule = arg.top_rule;
        arg_adj.sub_arguments = arg.sub_arguments;
        arg_adj.type = arg.type;
        arg_adj.strength = arg.strength;

        if (attack_index.count(arg.id)) {
            arg_adj.adj = attack_index[arg.id];
        }

        argumentadj.push_back(arg_adj);
    }
}

std::vector<Argument> Engine::compute_cycles() {
    int n = argumentadj.size();
    std::vector<bool> visited(n, false);
    std::vector<bool> recStack(n, false);

    std::unordered_map<std::string, int> id_to_index;
    for (int i = 0; i < n; i++) {
        id_to_index[argumentadj[i].id] = i;
    }

    for (int i = 0; i < n; i++) {
        if (!visited[i] && is_cycle(visited, recStack, i, argumentadj, id_to_index)) {
            std::vector<Argument> cycle_args;
            
            for (int j = 0; j < n; j++) {
                if (recStack[j]) {
                    cycle_args.push_back(arguments[j]); 
                }
            }
            return cycle_args;
        }
    }

    return {};
}

void Engine::compute_argument_strengths() {
    for (auto& arg : arguments) {
        if (arg.type=="atomic") {
            auto it = std::find_if(atoms.begin(), atoms.end(), [&arg](const Atom& atom) {
                return atom.id == arg.conclusion;
            });
            if (it != atoms.end()) {
                arg.strength = it->strength;
            }
        } else {
            double top_rule_strength = 0.0;
            auto it = std::find_if(rules.begin(), rules.end(), [&arg](const Rule& rule) {
                return rule.id == arg.top_rule.value_or("");
            });
            if (it != rules.end()) {
                top_rule_strength = it->strength;
            }
            double min_subarg_strength = 1.0;
            for (const auto& sub_arg_id : arg.sub_arguments) {
                auto sub_it = std::find_if(arguments.begin(), arguments.end(), [&sub_arg_id](const Argument& sub_arg) {
                    return sub_arg.id == sub_arg_id;
                });
                if (sub_it != arguments.end()) {
                    min_subarg_strength = std::min(min_subarg_strength, sub_it->strength);
                }
            }
            arg.strength = top_rule_strength * min_subarg_strength;
        }
        int_str[arg.id] = arg.strength;
    }
}

int Engine::propagate_strengths(double kappa, int iter, float epsilon){
    std::unordered_map<std::string, double> new_strength;
    if (iter>0) {
        for (int i = 0; i < iter; i++){
            for (auto& arg: arguments) {
                double alpha = 0.0;
                if (inv_attack_index.count(arg.id)) {
                    for (auto& att_arg: inv_attack_index[arg.id]) {
                        auto it = std::find_if(arguments.begin(), arguments.end(), [&att_arg](const Argument& index_arg) {
                            return index_arg.id == att_arg;
                        });
                        if (it != arguments.end()) {
                            alpha -= it->strength;
                        }
                    }
                }
                new_strength[arg.id] = std::clamp(int_str[arg.id] + kappa * std::tanh(alpha), 0.0, 1.0);
            }
            for (auto& arg : arguments) {
                arg.strength = new_strength[arg.id];
            }
        }
        return iter;
    } else {
        int safe_cap = 100000;
        float max_move = epsilon;
        while (max_move >= epsilon && safe_cap-- > 0) {
            max_move = 0.0;
            for (auto& arg: arguments) {
                double alpha = 0.0;
                if (inv_attack_index.count(arg.id)) {
                    for (auto& att_arg: inv_attack_index[arg.id]) {
                        auto it = std::find_if(arguments.begin(), arguments.end(), [&att_arg](const Argument& index_arg) {
                            return index_arg.id == att_arg;
                        });
                        if (it != arguments.end()) {
                            alpha -= it->strength;
                        }
                    }
                }
                float move = std::abs(arg.strength - (int_str[arg.id] + kappa * std::tanh(alpha)));
                if (move > max_move) max_move = move;
                new_strength[arg.id] = std::clamp(int_str[arg.id] + kappa * std::tanh(alpha), 0.0, 1.0);
            }
            for (auto& arg : arguments) {
                arg.strength = new_strength[arg.id];
            }
        }
        return 10000-safe_cap;
    }
}

File: PNsAV/src/core/engine.py
import json
import sys
import time
from pipeline import Log
import os
import platform

def main():
    mingw_bin_path = r"C:\\msys64\\mingw64\bin"

    if os.path.exists(mingw_bin_path):
        os.add_dll_directory(mingw_bin_path)
    else:
        print(f"Warning: MinGW path not found at {mingw_bin_path}. Check your installation.", file=sys.stderr)

    from engine import ArgEngine as module

    def topological_sort_arguments(arguments):
        id_to_arg = {arg["id"]: arg for arg in arguments}
        in_degree = {arg["id"]: 0 for arg in arguments}
        dependents = {arg["id"]: [] for arg in arguments}

        for arg in arguments:
            for sub_id in arg["sub_arguments"]:
                if sub_id not in id_to_arg:
                    raise ValueError(
                        f"Argument {arg['id']} references unknown subargument {sub_id}"
                    )
                dependents[sub_id].append(arg["id"])
                in_degree[arg["id"]] += 1

        ready = [aid for aid, deg in in_degree.items() if deg == 0]
        sorted_ids = []

        while ready:
            current = ready.pop()
            sorted_ids.append(current)
            for dep in dependents[current]:
                in_degree[dep] -= 1
                if in_degree[dep] == 0:
                    ready.append(dep)

        if len(sorted_ids) != len(arguments):
            raise ValueError(
                "Cycle detected in argument construction graph — "
                "this should never happen; check the validation layer."
            )

        return [id_to_arg[aid] for aid in sorted_ids]

    payload = json.loads(sys.stdin.read())

    atoms_payload = payload["atoms"]
    rules_payload = payload["rules"]
    args_payload = payload["args"]
    attacks_payload = payload["attacks"]
    prior_logs = payload.get("logs", [])
    kappa = payload["kappa"]
    epsilon = payload["epsilon"]
    iters = payload["iters"]

    logs = [Log(l[0], l[1]) for l in prior_logs]

    start = time.perf_counter()

    engine = module.Engine()

    for atom in atoms_payload:
        obj_atom = module.Atom()
        obj_atom.id = atom["id"]
        obj_atom.text = atom["text"]
        obj_atom.kb_type = atom["kb_type"]
        obj_atom.source_quote = atom["source_quote"]
        obj_atom.strength = atom["strength"]
        engine.add_atom(obj_atom)

    for rule in rules_payload:
        obj_rule = module.Rule()
        obj_rule.id = rule["id"]
        obj_rule.conclusion = rule["conclusion"]
        obj_rule.premises = rule["premises"]
        obj_rule.type = rule["type"]
        obj_rule.strength = rule["strength"]
        engine.add_rule(obj_rule)

    sorted_args = topological_sort_arguments(args_payload)

    for arg in sorted_args:
        obj_arg = module.Argument()
        obj_arg.id = arg["id"]
        obj_arg.conclusion = arg["conclusion"]
        obj_arg.top_rule = arg["top_rule"]
        obj_arg.sub_arguments = arg["sub_arguments"]
        obj_arg.type = arg["type"]
        obj_arg.strength = 1.0
        engine.add_argument(obj_arg)

    for uc in attacks_payload["undercutters"]:
        obj_attack = module.Attack()
        obj_attack.attacker = uc[0]
        obj_attack.target = uc[1]
        obj_attack.type = "undercutter"
        engine.add_attack(obj_attack)

    for rb in attacks_payload["rebuttals"]:
        obj_attack = module.Attack()
        obj_attack.attacker = rb[0]
        obj_attack.target = rb[1]
        obj_attack.type = "rebuttal"
        engine.add_attack(obj_attack)

    for um in attacks_payload["underminers"]:
        obj_attack = module.Attack()
        obj_attack.attacker = um[0]
        obj_attack.target = um[1]
        obj_attack.type = "underminer"
        engine.add_attack(obj_attack)

    engine.build_attack_map()
    engine.build_argumentadj_vector()

    cycles = engine.compute_cycles()
    engine.compute_argument_strengths()
    rounds = engine.propagate_strengths(kappa, iters, epsilon)

    logs.append(Log(f"Rounds executed until strength converged: {rounds}.", "info"))
    logs.append(Log(f"Found {len(cycles)} cycles in the graph.", "info"))
    logs.append(Log(f"Engine took {time.perf_counter()-start:.4f} seconds to run.", "info"))

    def print_atom(atom):
        print(atom.id, atom.kb_type, atom.text, atom.strength, atom.source_quote, sep="|", end="-")

    def print_rule(rule):
        print(rule.id, rule.premises, rule.conclusion, rule.type, rule.strength, sep="|", end="-")

    def print_arg(arg):
        print(arg.id, arg.type, arg.top_rule, arg.sub_arguments, arg.strength, arg.conclusion, sep="|", end="-")

    def print_attack(attack):
        print(attack.type, attack.target, attack.attacker, sep="|", end="-")

    def print_log(log):
        print(log.text, log.type, sep="|", end="-")

    for i in engine.atoms:
        print_atom(i)
    print("@", end="")
    for i in engine.rules:
        print_rule(i)
    print("@", end="")
    for i in engine.arguments:
        print_arg(i)
    print("@", end="")
    for i in engine.attacks:
        print_attack(i)
    print("@", end="")
    for i in logs:
        print_log(i)

if __name__ == "__main__":
    main()